api/xds_protocol.rst

```diff
@@ -36,43 +36,157 @@ occurs.
+   -  SotW: EndpointDiscoveryService.StreamEndpoints
+   -  Incremental: EndpointDiscoveryService.DeltaEndpoints
+-  Secret: Secret Discovery Service (SDS)
+   -  SotW: SecretDiscoveryService.StreamSecrets
+   -  Incremental: SecretDiscoveryService.DeltaSecrets
+-  Runtime: Runtime Discovery Service (RTDS)
+   -  SotW: RuntimeDiscoveryService.StreamRuntime
+   -  Incremental: RuntimeDiscoveryService.DeltaRuntime
+
+In the aggregated protocol variants, all resource types are multiplexed on a single gRPC stream, where each resource type is
+treated as a separate logical stream within the aggregated stream. In effect, it simply combines all of the above separate
+APIs into a single stream by treating requests and responses for each resource type as a separate sub-stream on the single
+aggregated stream. The RPC service and methods for the aggregated protocol variants are:
+
+-  SotW: AggregatedDiscoveryService.StreamAggregatedResources
+-  Incremental: AggregatedDiscoveryService.DeltaAggregatedResources
+
+For all of the SotW methods, the request type is :ref:`DiscoveryRequest <envoy_api_msg_DiscoveryRequest>` and the response
+type is :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`.
+
+For all of the incremental methods, the request type is :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>` and
+the response type is :ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`.
+
+Configuring Which Variant to Use
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the xDS API, the :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` message indicates how to obtain resources of a
+particular type. If the :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` contains a gRPC :ref:`ApiConfigSource
+<envoy_api_msg_core.ApiConfigSource>`, it points to an upstream cluster for the management server; this will initiate an
+independent bidirectional gRPC stream for each xDS resource type, potentially to distinct management servers. If the
+:ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` contains a :ref:`AggregatedConfigSource
+<envoy_api_msg_core.AggregatedConfigSource>`, it tells the client to use :ref:`ADS <xds_protocol_ads>`.
+
+Currently, the client is expected to be given some local configuration that tells it how to obtain the :ref:`Listener
+<envoy_api_msg_Listener>` and :ref:`Cluster <envoy_api_msg_Cluster>` resources. :ref:`Listener <envoy_api_msg_Listener>`
+resources may include a :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` that indicates how the :ref:`RouteConfiguration
+<envoy_api_msg_RouteConfiguration>` resources are obtained, and :ref:`Cluster <envoy_api_msg_Cluster>` resources may include a
+:ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` that indicates how the :ref:`ClusterLoadAssignment
+<envoy_api_msg_ClusterLoadAssignment>` resources are obtained.
+
+In Envoy, the bootstrap file contains two :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` messages,
```

Comment:
As we potentially grow more clients of xDS/UDPA, I'm wondering if we should format this document to have global information and then potentially sub-sections for client specific information? This is mainly a readability thing around having some kind of header, etc. to convey client specific information. WDYT? I think it could be as simple as a sub-heading at the end of a section?

Reply:
I had originally tried to focus this doc on the xDS transport protocol itself, without details on individual clients.  But in response to an earlier comment by @htuch (see https://github.com/envoyproxy/envoy/pull/9029#discussion_r346871814), I had to change this to talk about the individual clients.

I completely agree with the sentiment that we should be able to pull apart the transport protocol from the way individual clients are configured to use it.  In the long term, probably as part of the UDPA effort, I would like to see a formal spec for the transport-protocol that does not need to contain any client-specific configuration details -- i.e., the client details would not just be in a separate section but would actually be in completely separate docs.  But unfortunately, where we are today, it's really hard to understand the API without understanding how the clients use it, so I think it would actually hurt readability somewhat to try to separate this out.

Reply:
I was mostly suggesting to just add a header of some kind of to make it more clear to the reader when we are entering client specific logic vs. actually splitting it out into a separate doc.

Reply:
Oh, okay, that's easy enough.  Done.